### ISSUE TITLE
retry not working with POST (with OPTIONS preflight) and 401 statuscode

### ISSUE THREAD

**Issue by DanielRuf** [sindresorhus/ky#230]
We have the following retry options:

```
const retryOptions = {
        limit: 10,
        methods: ['post'],
        statusCodes: [401, 500, 502]
};
```

I generate a 401 error from the backend and I see it also as 401 statuscode in the developer tools of the browser.

But it does not retry the failing request:

```
const response = await ky.post(`${API_ENDPOINT}/users`, {
            json: {
                accessToken: accessToken
            },
            retry: retryOptions
        }).json();
        return response;
```

Any idea why?
The docs are not very helpful here.

Is `post` not supported there?

**Comment by DanielRuf** [comment]
Ah, some debugging in the node_modules/ky/index.js reveals:

`retry error TypeError: "Body has already been consumed."`

**Comment by DanielRuf** [comment]
I think we have to do the `json()` call in a separate line.

**Comment by DanielRuf** [comment]
Hm, now it is `retry error HTTPError: "Unauthorized"` and after that the `retry error TypeError: "Body has already been consumed."`

**Comment by DanielRuf** [comment]
This thrown be the last line here:

```
if (this._options.timeout === false) {
      return globals.fetch(this.request);
    }

    return timeout(globals.fetch(this.request), this._options.timeout, this.abortController);
```

**Comment by DanielRuf** [comment]
So for me it seems POST is not suppported right now.

**Comment by sholladay** [comment]
Duplicate of #217 

Hi @DanielRuf, thanks for the bug report. Your use case is officially supported, however `v0.16` has a regression related to constructing the request that prevents it from working. This is a known issue.

We need to fix Ky by calling `this.request.clone()` everywhere we try to copy or use the request, which I plan to implement soon (but feel free to make a PR if you need this fixed quickly).

Until such a fix is released, if this feature is important to you, you may want to stay on `v0.15`.

**Comment by sholladay** [comment]
Hi @DanielRuf, thanks for the bug report. Your use case is officially supported, however `v0.16` has a regression related to constructing the request that prevents it from working. This is a known issue.

We need to fix Ky by calling `this.request.clone()` everywhere we try to copy or use the request, which I plan to implement soon (but feel free to make a PR if you need this fixed quickly).

Until such a fix is released, if this feature is important to you, you may want to stay on `v0.15`.

**Comment by sholladay** [comment]
Duplicate of #217

**Comment by DanielRuf** [comment]
We've migrated to simple get requests and moved the body to custom headers which works without any issues as this issue is a blocker for us.

**Comment by DanielRuf** [comment]
I do not see where it is related to #217 as we have different errors which are only because of the POST body. Or is this what is also caused by the regression and solved by cloning the request?

Because this is not clear in #217 =)

**Comment by sholladay** [comment]
You mentioned seeing `TypeError: "Body has already been consumed."`. This is the error that #217 is all about, hence why I marked this as a duplicate. I think it is the root cause of your issue, since it prevents retries from working when the request has a body.

Retries should still work with any request that does not have a body, including `POST` requests (assuming you opt-in to retrying `POST`).

**Comment by DanielRuf** [comment]
> You mentioned seeing `TypeError: "Body has already been consumed."`. This is the error that #217 is all about,



> I do not see where it is related to #217

A simple `ctrl f` does not show any result for `Body has already been consumed`, that's why I am asking ;-) Even no result for `consumed` and `body`.

**Comment by sholladay** [comment]
You are using Firefox whereas they are using Chrome, so the message is slightly different... sorry it's not standardized. 😄 

> TypeError: "Body has already been consumed."

vs

> TypeError: Failed to execute 'fetch' on 'Window': Cannot construct a Request with a Request object that has already been used.

These messages are for the same error condition, here's a minimal repro that you can run in the console of both browsers to prove it:

```js
const req = new Request('/foo', { method : 'POST', body : 'foo' });
fetch(req);
fetch(req);
```

I just submitted a fix for this in PR #231. You can try using that branch and check that it does indeed work for your use case (you'll have to load the dependency from GitHub, which you can do with npm or by `import`ing the raw GitHub URL in the browser).

**Comment by DanielRuf** [comment]
Ah, thanks for the clarification. That makes sense now =)
Sigh, I hope we get consistent error messages in browsers :D 

Or people can try it with `patch-package` and use https://github.com/sindresorhus/ky/pull/231.patch ;-)
